Frameworks/CoreGraphics/CGPath.mm

```diff
@@ -31,8 +31,92 @@
 #import <algorithm>
 #import <vector>
 
+#import <CoreFoundation/CoreFoundation.h>
+#import <CFRuntime.h>
+#import "D2DWrapper.h"
+
+#include <COMIncludes.h>
+#import <wrl/client.h>
+#import <D2d1.h>
+#include <COMIncludes_End.h>
+
 static const wchar_t* TAG = L"CGPath";
 
+using namespace std;
+using namespace Microsoft::WRL;
+
+struct __CGPath {
+    CFRuntimeBase _base;
+
+    ComPtr<ID2D1PathGeometry> _pathGeometry;
+    ComPtr<ID2D1GeometrySink> _geometrySink;
+
+    bool _isPathClosed;
+    CGPoint _currentPoint;
+    CGPoint _startingPoint;
+};
+
+// A private helper function for re-opening a path geometry.
+static void __preparePathForEditing(CGPathRef path) {
+    if (path->_isPathClosed) {
+        // Re-open this geometry.
+        ComPtr<ID2D1Factory> factory = _CreateD2DFactoryInstance();
+        if (FAILED(factory->CreatePathGeometry(&path->_pathGeometry))) {
+            return;
+        }
+        if (FAILED(path->_pathGeometry->Open(&path->_geometrySink))) {
+            return;
+        }
+
+        // This path is no longer closed.
+        path->_isPathClosed = false;
+    }
+}
+
+static void __closePath(CGPathRef path) {
+    if (!path->_isPathClosed) {
+        path->_geometrySink->Close();
+        path->_isPathClosed = true;
+    }
+}
+
+static Boolean __CGPathEqual(CFTypeRef cf1, CFTypeRef cf2) {
+    struct __CGPath* path1 = (struct __CGPath*)cf1;
+    struct __CGPath* path2 = (struct __CGPath*)cf2;
+
+    return true;
+}
+
+static void __CGPathInit(CFTypeRef cf) {
+    CGPathRef path = (CGPathRef)cf;
+    // ComPtr needs to be manually new'd, to avoid calling Create functions to an incomplete out-pointer
+    struct __CGPath* mutablePath = const_cast<struct __CGPath*>(path);
+    new (std::addressof(mutablePath->_pathGeometry)) ComPtr<ID2D1PathGeometry>();
+    new (std::addressof(mutablePath->_geometrySink)) ComPtr<ID2D1GeometrySink>();
```

Comment:
You already did this.
Perhaps it can be done in init for clarity's sake!

---
In reply to: [83128234](https://github.com/Microsoft/WinObjC/pull/1138#discussion_r83128234) [](ancestors = 83128234)